### PR TITLE
RMP-12464 Integrate FreeIPA support

### PR DESCRIPTION
**Pull request by lacikaaa**

### CODE REVIEW

File: core-model/src/main/java/com/sequenceiq/cloudbreak/domain/KerberosConfig.java [hunk 1]
```diff
@@ -72,6 +72,12 @@
     @SecretValue
     private Secret krb5Conf = Secret.EMPTY;
 
+    @Column(name = "domain")
```
Comment (horadla23) [inline core-model/src/main/java/com/sequenceiq/cloudbreak/domain/KerberosConfig.java]: i think this is not needed

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/model/KerberosBase.java [hunk 2]
```diff
@@ -47,9 +41,17 @@
...
     private Boolean tcpAllowed = false;
 
-    @NotNull
+    @NotNull(message = "Kerberos type can not be null")
     private KerberosType type;
 
+    @ApiModelProperty(StackModelDescription.KERBEROS_DOMAIN)
+    private String domain;
+
+    @Pattern(regexp = "(^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(,((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}"
```
Comment (gregito) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/model/KerberosBase.java]: please write some test for this regex. not just for breaking the test if someone changes the pattern, but to provide examples through the test also.
(like in BlueprintRequestTest.java)

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/KerberosRequestToKerberosConfigConverter.java [hunk 3]
```diff
@@ -25,6 +26,12 @@ public KerberosConfig convert(KerberosRequest source) {
         kerberosConfig.setContainerDn(source.getContainerDn());
         kerberosConfig.setDescriptor(source.getDescriptor());
         kerberosConfig.setKrb5Conf(source.getKrb5Conf());
+        if (StringUtils.isNotBlank(source.getDomain())) {
```
Comment (gregito) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/KerberosRequestToKerberosConfigConverter.java]: please write unit test for this converter

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCreationSetupService.java [hunk 4]
```diff
@@ -214,6 +217,12 @@ public Cluster prepare(ClusterRequest request, Stack stack, Blueprint blueprint,
         return savedCluster;
     }
 
+    private void decorateStackWithCustomDomainIfAdOrIpaJoinable(Stack stack, Cluster cluster) {
```
Comment (gregito) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCreationSetupService.java]: unit test, please. the calling method is much larger but writing a dedicated test case which tests this case would be more than nothing.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/PreTerminationStateExecutor.java [hunk 5]
```diff
@@ -31,13 +32,23 @@
     @Inject
     private StackUtil stackUtil;
 
-    public void leaveAdDomain(Stack stack) throws CloudbreakException {
-        if (stack.getCluster().isAdJoinable()) {
+    public void runPreteraminationTasks(Stack stack) throws CloudbreakException {
+        leaveDomains(stack);
+    }
+
+    private void leaveDomains(Stack stack) throws CloudbreakException {
```
Comment (gregito) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/PreTerminationStateExecutor.java]: unit tests, please.

File: template-manager-blueprint/src/main/java/com/sequenceiq/cloudbreak/blueprint/kerberos/KerberosBlueprintService.java [hunk 6]
```diff
@@ -98,7 +102,11 @@ public BlueprintTextProcessor extendBlueprintWithKerberos(BlueprintTextProcessor
             SiteConfigurations configs = SiteConfigurations.getEmptyConfiguration();
             Map<String, String> krb5Conf = new HashMap<>();
             krb5Conf.put("domains", domains);
-            krb5Conf.put("manage_krb5_conf", "true");
+            if (kerberosType != KerberosType.EXISTING_AD && kerberosType != KerberosType.EXISTING_FREEIPA) {
```
Comment (gregito) [inline template-manager-blueprint/src/main/java/com/sequenceiq/cloudbreak/blueprint/kerberos/KerberosBlueprintService.java]: do we have blueprint module test case for such an event like this?

File: template-manager-blueprint/src/main/java/com/sequenceiq/cloudbreak/blueprint/kerberos/KerberosDetailService.java [hunk 7]
```diff
@@ -24,10 +24,14 @@
     private final Gson gson = new Gson();
 
     public String resolveTypeForKerberos(@Nonnull KerberosConfig kerberosConfig) {
```
Comment (gregito) [inline template-manager-blueprint/src/main/java/com/sequenceiq/cloudbreak/blueprint/kerberos/KerberosDetailService.java]: there is no unit test neither the change or the method

**Pull request merged**